packages/terra-aggregator/src/Aggregator.jsx

```diff
@@ -0,0 +1,229 @@
+
+      if (!focusItemIsPresent) {
+        this.releaseFocus(undefined, true);
+      }
+    }
+  }
+
+  setFocusState(itemKey, itemState) {
+    this.setState({
+      focusedItemId: itemKey,
+      focusedItemState: itemState,
+    });
+  }
+
+  resetFocusState() {
+    this.setFocusState();
+  }
+
+  requestFocus(itemId, itemState) {
+    const { disclose } = this.props;
+    const { focusedItemId } = this.state;
+
+    return Promise.resolve()
+      .then(() =>
+        /**
+         * Focus is released on the currently focused item to ensure a clean start for the next component receiving focus.
+         * The releaseFocus's Promise is returned and inserted into the Promise chain to prevent disclosures from occurring
+         * if the focus release fails.
+         */
+        this.releaseFocus(focusedItemId)
+          .then(() => {
+            this.setFocusState(itemId, Object.freeze(itemState || {}));
+          }),
+      )
+      .then(() => {
+        const focusRequestPayload = {};
+
+        /**
+         * If the Aggregator is provided with disclosure functionality, the focus request is resolved with a custom
+         * disclose implementation.
+         */
+        if (disclose) {
+          focusRequestPayload.disclose = (data) => {
+            /**
+             * If the itemId no longer matches the focusedItemId, then the disclose is being called after
+             * the item has lost focus. This can happen if an Aggregator item caches the disclosue function they're given
+             * and calls it later.
+             */
+            if (this.state.focusedItemId !== itemId) {
+              return Promise.reject();
+            }
+
+            return disclose(data)
+              .then(({ afterDismiss, dismissDisclosure, ...other }) => {
+                /**
+                 * The disclosure's dismissDisclosure instance is cached so it can be called later. If an Aggregator item is
+                 * currently presenting a disclosure and releases focus, we will call this function to force
+                 * the disclosure to close.
+                 */
+                this.forceDismissInstance = dismissDisclosure;
+
+                /**
+                 * A handler is added to the deferred afterDismiss promise chain to remove the cached dismissDisclosure instance (the disclosure is
+                 * closing, so it is no longer relevant). The handler also resets the focus state if focus is currently held by a component.
+                 */
+                afterDismiss.then(() => {
+                  this.forceDismissInstance = undefined;
+
+                  if (this.state.focusedItemId) {
+                    this.resetFocusState();
+                  }
+                });
+
+                // We return the same API so as not to disrupt the chain.
+                return { afterDismiss, dismissDisclosure, ...other };
+              });
+          };
+        }
+
+        return focusRequestPayload;
```

Comment:
What's the use case for this?

Reply:
This is how components withing the Aggregator get access to the disclose function. They will not receive a traditional `app` prop (not from the Aggregator, at least), because we're trying to ensure that only the focused item has access to the progressive disclosure API.

This is what it'd look like from the item's standpoint.
```jsx
aggregatorDelegate.requestFocus({...})
  .then((focusRequestPayload) => {
    focusRequestPayload.disclose({...});
  });
```

Reply:
whoops read this wrong. 👍